### DISCUSSION TITLE
foreign_keys and _user_defined_foreign_keys attributes not found on relationship property

### DISCUSSION THREAD

**Discussion by abhay-apella** [sqlalchemy/sqlalchemy#11286] [Usage Questions]
I am getting `AttributeError: _user_defined_foreign_keys` on relationship fields in my versioned table.

Table one is a type table:
```
class Base(AsyncAttrs, DeclarativeBase):
    pass

class NotificationType(Base, Versioned)
    __tablename__ = "notification_types"
    id: Mapped[str] = mapped_column(String, primary_key=True)
    name: Mapped[str] = mapped_column(String, nullable=False)

class Notification(Base)
    __tablename__ = "notification"
    id: Mapped[str] = mapped_column(String, primary_key=True)
    ...
    notification_type_id: Mapped[str] = mapped_column(ForeignKey("notification_types.id"), nullable=False, index=True)
    notification_relationship: Mapped["NotificationType"] = relationship(
        NotificationType, back_populates="NotificationType", foreign_keys=["notification_type_id"], lazy="joined", innerjoin=True
    )
```
In my Versioned mixin definition (for history tables) I have mostly followed the code listed [here](https://github.com/sqlalchemy/sqlalchemy/blob/main/examples/versioned_history/history_meta.py).

I am also looping over the relationship attributes in order to try to carry over the notification type and relationship information:
```
def _history_mapper(local_mapper):
    if not super_mapper or local_mapper.local_table is not super_mapper.local_table:
        ...
        for prop in local_mapper.iterate_properties:
            if isinstance(prop, RelationshipProperty):
                columns = [fk.key for fk in prop._user_defined_foreign_keys]
        ...
```
I notice that when attempting to run this code, building fails on line columns = [fk.key for fk in prop._user_defined_foreign_keys] as _user_defined_foreign_keys is undefined. It will sometimes work if I debug and wait some amount of time at that line, otherwise I will get an attribute error:
```
    def _fallback_getattr(self, key):
        raise AttributeError(key)
->
    E   AttributeError: _user_defined_foreign_keys
```
I am using SQLAlchemy = {version = "2.0.28", extras = ["asyncio"]} .

I do not see any relevant documentation about why foreign_keys and _user_defined_foreign_keys attributes being lazy loaded during build or being awaitable. This behavior worked in SqlAlchemy 1.4 and I also do not see any documentation updates on this behavior changing (without `AsyncAttrs`). Is this because the Base class extends AsyncAttrs and needs to be awaited? Or am I doing something else wrong?

**Comment by CaselIT** [comment]
Hi,

You are accessing a private attribute, so any change to it would not be documented.
What are you trying to do with that argument? The linked version history example makes no use of that attribute

**Reply by abhay-apella** [reply]
There is some mention of it in this section: https://docs.sqlalchemy.org/en/14/orm/basic_relationships.html#late-evaluation-of-relationship-arguments. As to the actual use of the variable I am not sure. The behavior I am trying to transfer over from v1.4 to v2 is that versioned tables should copy over relationships to the history and there is some custom logic with relationships and foreign keys. In the same vein I cannot access other attributes of the relationship either such as `foreign_keys`, though I can see other attributes are present on the property like `lazy` or  `innerjoin` when I debug. Also, one weird behavior is that `foreign_keys` and `_user_defined_foreign_keys` are eventually accessible in my debugger session if I just pause at the access for some random amount of time per table/relationship.

**Reply by CaselIT** [reply]
That property is lazily initialized, so when you are accessing that the replationship is probably not yet inited.

I would suggest doing something similar to what versioned does in the example to ensure that you call your logic after the target class has finished configuring

**Reply by abhay-apella** [reply]
My logic only lives in the `_history_mapper` function, and much of it (save for my additions) does match the code in the example. If it is being lazily loaded, what would I wait on to make sure this additional logic in the Versioned mixin is called after the class has finished configuring?

**Reply by abhay-apella** [reply]
I also took a shot at trying to access the property fields like `foreign_keys` after events such as in `create_version` or `before_flush` and it does not seem to be initialized. Is this initialization a function of build hierarchy?

**Reply by CaselIT** [reply]
you can try forcing the initialization with https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.configure_mappers

sorry but I haven't had time to test this yet. will try to provide additional suggestion once I do

**Reply by abhay-apella** [reply]
That kind of works! If I add add 
```
from sqlalchemy.orm import attributes, object_mapper, configure_mappers
...
def _history_mapper(local_mapper):
    configure_mappers()
```

The field `_user_defined_foreign_keys` is no longer null though the `foreign_keys` attribute is still null. Though I assume this function wasn't intended to be called manually. I think the old code I am working on was written with the intent of mapped imperatively and transitioning to 2.0 breaks that. In order to limit the amount of changes I would have to do, would it be better to remove the Versioned mixin from my model objects and instead add a [mapper.configured](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.MapperEvents.mapper_configured) listener that then sets the history table?
